### PR TITLE
[Graphbolt] Change input/output  of sample neighbors to hetero format

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Change the input and output of `sample_neighbors` to heterogeneous format.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 1]
```diff
@@ -213,13 +215,130 @@ def in_subgraph(self, nodes: torch.Tensor) -> torch.ScriptObject:
...
+        subgraph to general struct 'SampledSubgraphImpl'."""
+        column_num = (
+            C_sampled_subgraph.indptr[1:] - C_sampled_subgraph.indptr[:-1]
+        )
+        column = C_sampled_subgraph.reverse_column_node_ids.repeat_interleave(
+            column_num
+        )
+        row = C_sampled_subgraph.indices
+        type_per_edge = C_sampled_subgraph.type_per_edge
+        if type_per_edge is None:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:232]:
```suggestion
        if type_per_edge is None:
            # The sampled graph is already a homogeneous graph.
```

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 2]
```diff
@@ -213,13 +215,130 @@ def in_subgraph(self, nodes: torch.Tensor) -> torch.ScriptObject:
...
+            node_pairs = (row, column)
+        else:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:234]:
```suggestion
        else:
            # The sampled graph is a fused homogenized graph, which need to be
            # converted to heterogeneous graphs.
```

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 3]
```diff
@@ -213,13 +215,130 @@ def in_subgraph(self, nodes: torch.Tensor) -> torch.ScriptObject:
...
+                dst_ntype_id = self.metadata.node_type_to_id[dst_ntype]
+                mask = type_per_edge == etype_id
+                hetero_row = row[mask] - self.node_type_offset[src_ntype_id]
+                hetero_column = (
+                    column[mask] - self.node_type_offset[dst_ntype_id]
+                )
+                node_pairs[etype] = (hetero_row, hetero_column)
+        return SampledSubgraphImpl(
+            node_pairs=node_pairs,
+        )
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:248]:
```suggestion
        return SampledSubgraphImpl(node_pairs=node_pairs,)
```

**Review by frozenbugs — COMMENTED** [review]
LGTM overall, need to double check after the diffbase is merged.

File: python/dgl/graphbolt/impl/csc_sampling_graph.py (hunk 1 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/csc_sampling_graph.py:232]: added

File: python/dgl/graphbolt/impl/csc_sampling_graph.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/csc_sampling_graph.py:248]: fixed

File: python/dgl/graphbolt/impl/csc_sampling_graph.py (hunk 2 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/csc_sampling_graph.py:234]: added

File: python/dgl/graphbolt/impl/csc_sampling_graph.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/csc_sampling_graph.py:248]: Use `return SampledSubgraphImpl(node_pairs=node_pairs) ` instead.

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py [hunk 4]
```diff
@@ -379,127 +379,211 @@ def test_sample_neighbors():
...
+        subgraph.node_pairs[("n1", "e1", "n2")][0].numel()
+        == expected_sampled_num1
+    )
+    assert (
+        subgraph.node_pairs[("n2", "e2", "n1")][0].numel()
+        == expected_sampled_num2
+    )
 
 
+# test_sample_neighbors_fanouts([4], 4)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:530]: remove?

**Review by frozenbugs — APPROVED** [review]

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py (hunk 4 above)
Comment (peizhou001) [reply tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:530]: removed

**Pull request merged**